---
title: "Hazan_ex4"
author: "Josh Hazan"
date: "11/18/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#### Required packages and loading data

```{r packages}
library(tidyverse)
library(compGenomRData)
library(DESeq2)
library(gProfileR)
library(gage)
library(knitr)

counts_file <- system.file("extdata/rna-seq/SRP029880.raw_counts.tsv",
package = "compGenomRData")
coldata_file <- system.file("extdata/rna-seq/SRP029880.colData.tsv",
package = "compGenomRData")
counts <- as.matrix(read.table(counts_file, header = T, sep = '\t'))

#remove the 'width' column
countData <- as.matrix(subset(counts, select = c(-width)))
#define the experimental setup
colData <- read.table(coldata_file, header = T, sep = '\t',
stringsAsFactors = TRUE)
#define the design formula
designFormula <- "~ group"
#create a DESeq dataset object from the count matrix and the colData
dds <- DESeqDataSetFromMatrix(countData = countData,
                              colData = colData,
                              design = as.formula(designFormula))
dds <- DESeq(dds)
DEresults = results(dds, contrast = c("group", 'CASE', 'CTRL'))

##Select genes significantly differentially expressed between case and ctrl

#remove genes with NA values
DE <- DEresults[!is.na(DEresults$padj),]
#select genes with adjusted p-values below 0.1
DE <- DE[DE$padj < 0.1,]
#select genes with absolute log2 fold change above 1 (two-fold change)
DE <- DE[abs(DE$log2FoldChange) > 1,]
#get the list of genes of interest
genesOfInterest <- rownames(DE)
#calculate enriched GO terms
goResults <- gprofiler(query = genesOfInterest,
                       organism = 'hsapiens',
                       src_filter = 'GO',
                       hier_filtering = 'moderate')
```

### Question 1
```{r}
goResults.q1 <- gprofiler(query = genesOfInterest,
                       organism = 'hsapiens',
                       src_filter = c("GO","KEGG","CORUM","REACTOME"),
                       hier_filtering = 'moderate')
goResults.q1 <- goResults.q1[order(goResults.q1$recall),]
```

## Gene set enrichment analysis
```{r preparing data}
#Let's define the first gene set as the list of genes from one of the
#significant GO terms found in the GO analysis. order go results by pvalue
goResults <- goResults[order(goResults$p.value),]
#restrict the terms that have at most 100 genes overlapping with the query
go <- goResults[goResults$overlap.size < 100,]
# use the top term from this table to create a gene set
geneSet1 <- unlist(strsplit(go[1,]$intersection, ','))
#Define another gene set by just randomly selecting 25 genes from the counts
#table get normalized counts from DESeq2 results
normalizedCounts <- DESeq2::counts(dds, normalized = TRUE)
geneSet2 <- sample(rownames(normalizedCounts), 25)
geneSets <- list('top_GO_term' = geneSet1,
                 'random_set' = geneSet2)
# Using the defined gene sets, weâ€™d like to do a group comparison between the case
# samples with respect to the control samples.
#Use the normalized counts to carry out a GSEA.
gseaResults <- gage(exprs = log2(normalizedCounts+1),
                    ref = match(rownames(colData[colData$group =='CTRL',]),
                                colnames(normalizedCounts)),
                    samp = match(rownames(colData[colData$group == 'CASE',]),
                                 colnames(normalizedCounts)),
                    gsets = geneSets, compare = 'as.group')
gseaResults$greater
gseaResults$less
```

### Question 2
```{r}
#Using "paired", the default compare scheme
gseaResults2 <- gage(exprs = log2(normalizedCounts+1),
                    ref = match(rownames(colData[colData$group =='CTRL',]),
                                colnames(normalizedCounts)),
                    samp = match(rownames(colData[colData$group == 'CASE',]),
                                 colnames(normalizedCounts)),
                    gsets = geneSets, compare = 'paired')
gseaResults2$greater
gseaResults2$less

#Using "1ongroup".
gseaResults3 <- gage(exprs = log2(normalizedCounts+1),
                    ref = match(rownames(colData[colData$group =='CTRL',]),
                                colnames(normalizedCounts)),
                    samp = match(rownames(colData[colData$group == 'CASE',]),
                                 colnames(normalizedCounts)),
                    gsets = geneSets, compare = '1ongroup')
gseaResults3$greater
gseaResults3$less

#All show significant for greater, but not less in the top term column. However, paired and unpaired use different assumptions and therefore have different assumptions and also compare each case individually to the controls, rather than the groups as a whole.
#The 1ongroup command compares 1 sample at a time to the controls, and also gives a different result.
```

### Question 3
```{r}
goResults.q3 <- gprofiler(query = genesOfInterest,
                       organism = 'hsapiens',
                       src_filter = 'GO', significant = FALSE,
                       hier_filtering = 'moderate')
# Only 8 results were obtained, compared to 93 without the significant = FALSE argument.

ggplot(data = goResults.q3, aes(x = term.size, y = -log10(p.value))) +
  geom_point(aes(color = significant)) + theme_classic() + 
  geom_text(data = subset(goResults.q3, significant == "TRUE"), 
            aes(label = term.name), nudge_y = (-4))
# A lower p-value is associated with a higher term size
```

### Question 4
```{r}
# When doing all 10 together, the analysis returned with NA.
# Possibly because of too many gene names. Therefore, I am doing 1-5 and then 6-10.
geneSet.top1_5 <- unlist(strsplit(go[1:5,]$intersection, ','))
geneSet.top6_10 <- unlist(strsplit(go[6:10,]$intersection, ','))
geneSetlist.top10 <- list("GO_Terms_1_to_5" = geneSet.top1_5,
                          "GO_Terms_6_to_10" = geneSet.top6_10)
gseaResults.top10 <- gage(exprs = log2(normalizedCounts+1),
                    ref = match(rownames(colData[colData$group =='CTRL',]),
                                colnames(normalizedCounts)),
                    samp = match(rownames(colData[colData$group == 'CASE',]),
                                 colnames(normalizedCounts)),
                    gsets = geneSetlist.top10, compare = 'as.group')
head(gseaResults.top10$greater)
head(gseaResults.top10$less)

#Significance was seen for greater, not less

```

### Question 5




