---
title: "Hazan_ex3"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(DESeq2)
library(EnhancedVolcano)
library(compGenomRData)
library(edgeR)
library(compcodeR)
```

## Required files
```{r cars}
counts_file <- system.file("extdata/rna-seq/SRP029880.raw_counts.tsv",
package = "compGenomRData")
coldata_file <- system.file("extdata/rna-seq/SRP029880.colData.tsv",
package = "compGenomRData")

counts <- read.table(counts_file, sep = "\t", header = TRUE)
counts <- as.matrix(subset(counts, select = c(-width)))
coldata <- read.table(coldata_file, sep = "\t", header = TRUE)

counts.deseq <- DESeqDataSetFromMatrix(countData = counts,
                                       colData = coldata,
                                       design = ~ group)
counts.deseq <- counts.deseq[ rowSums(DESeq2::counts(counts.deseq)) >= 10, ]
counts.deseq <- DESeq(counts.deseq)
DE.results <- results(counts.deseq, contrast = c("group", 'CASE', 'CTRL'))
DE.results <- DE.results[order(DE.results$pvalue),]
DE.results.DF <- as.data.frame(DE.results)
```

## Question 1
```{r}
EnhancedVolcano(DE.results,
                lab = NA,x = 'log2FoldChange',y = 'pvalue',FCcutoff=1)

ggplot() + theme_classic() +
  geom_point(data = DE.results.DF, aes(x = log2FoldChange, y = -log10(padj)),
             col = "grey80", size = 1) + 
  geom_point(data = subset(DE.results.DF, log2FoldChange > 1 & padj < 0.05),
             col = "red2", size = 1, 
             aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(data = subset(DE.results.DF, log2FoldChange < (-1) & padj < 0.05),
             col = "steelblue2", size = 1, 
             aes(x = log2FoldChange, y = -log10(padj))) +
  theme(legend.title = element_blank()) + labs(x = "log2(fold change) case vs. control",
                                               y = "-log10(adjusted p-value)")
```

## Question 2
```{r}
plotDispEsts(counts.deseq, 
             genecol = "black", fitcol = "orange", finalcol = "purple",
             legend = TRUE,log = "xy",cex = 0.45)
```

## Question 3
```{r}
results.lfc1 <- DESeq2::results(counts.deseq, lfcThreshold = 1, pAdjustMethod = "BH")
results.lfc1 <- as.data.frame(results.lfc1)
# The default is 0, which indicates a fold change of 1. 
#Changing the threshold to 1 (i.e. fold change of 2) makes the cutoff for significance
#much more strict, and fewer hits were obtained, as shown in the below plot

ggplot() + theme_classic() +
  geom_point(data =results.lfc1, aes(x = log2FoldChange, y = -log10(padj)),
             col = "grey80", size = 1) + 
  geom_point(data = subset(results.lfc1, log2FoldChange > 1 & padj < 0.05),
             col = "red2", size = 1, 
             aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(data = subset(results.lfc1, log2FoldChange < (-1) & padj < 0.05),
             col = "steelblue2", size = 1, 
             aes(x = log2FoldChange, y = -log10(padj))) +
  theme(legend.title = element_blank()) + labs(x = "log2(fold change) case vs. control",
                                               y = "-log10(adjusted p-value)")
```

## Question 4
```{r}
# Independent filtering is filtering with custom cutoffs to ensure data with low reads/low reliability is not included by accident. This can vary depending on the data. 
```

## Question 5
```{r}
counts.edgeR <- counts
edgeR.list <- DGEList(counts = counts.edgeR, group = coldata$group)
filter.edgeR <- filterByExpr(edgeR.list)
edgeR.list <- edgeR.list[filter.edgeR,,keep.lib.sizes=FALSE]
edgeR.list <- calcNormFactors(edgeR.list)
design <- model.matrix(~coldata$group)
edgeR.list <- estimateDisp(edgeR.list,design)
edgeR.list <- exactTest(edgeR.list)
edgeR.list.df <- as.data.frame(edgeR.list)

#Volcano plot showing the difference. DESeq2 had ~1000 hits more than edgeR
ggplot() + theme_classic() +
  geom_point(data =edgeR.list.df, aes(x = logFC, y = -log10(PValue)),
             col = "grey80", size = 1) + 
  geom_point(data = subset(edgeR.list.df, logFC > 1 & PValue < 0.05),
             col = "red2", size = 1, 
             aes(x = logFC, y = -log10(PValue))) +
  geom_point(data = subset(edgeR.list.df, logFC < (-1) & PValue < 0.05),
             col = "steelblue2", size = 1, 
             aes(x = logFC, y = -log10(PValue))) +
  theme(legend.title = element_blank()) + labs(x = "log2(fold change) case vs. control",
                                               y = "-log10(adjusted p-value)")
```

## Question 6
```{r}
#I did not have time to try this question
```











